webui/react/src/modal/useModality.tsx

```diff
@@ -0,0 +1,183 @@
+  children: ReactNode;
+}
+
+function ModalContainer({ modalIsOpen, children: modalBody, setModalIsOpen }: ModalContainerProps) {
+  /*
+   * the params are defined here, but set from with the ModalBodyComponent
+   * via useModalParams (which uses React Context)
+   */
+  const [params, setParams] = useState<UseModalParams | null>(null);
+  const [isSubmitting, setIsSubmitting] = useState(false);
+
+  const close = useCallback(() => setModalIsOpen(false), [setModalIsOpen]);
+
+  const handleOk = useCallback(async () => {
+    setIsSubmitting(true);
+    await params?.submit?.handler();
+    setIsSubmitting(false);
+    setModalIsOpen(false);
+    await params?.submit?.onComplete?.();
+  }, [params?.submit, setModalIsOpen]);
+
+  if (params === null) {
+    /**
+     * we still want to render the modalBody, although invisibly,
+     * so that it can set the params (which will then allow it to be
+     * displayed)
+     */
+    return (
+      <ModalContext.Provider value={{ setParams }}>
+        <div style={{ display: 'none' }}>{modalBody}</div>
+      </ModalContext.Provider>
+    );
+  }
+
+  const { titleText, cancelText, submit, danger, icon, footerLink, headerLink, size } = params;
+  return (
+    <ModalContext.Provider value={{ params, setParams }}>
+      <AntdModal
+        cancelText={params.cancelText}
+        footer={
+          <div style={{ display: 'flex', justifyContent: 'space-between' }}>
+            <div>
+              {footerLink && (
+                <Link path={footerLink.url} popout>
+                  {footerLink.text}
+                </Link>
+              )}
+            </div>
+            <div>
+              <Button key="back" onClick={close}>
+                {cancelText}
+              </Button>
+              <Button
+                danger={danger}
+                disabled={!!submit?.disabled}
+                key="submit"
+                loading={isSubmitting}
+                type="primary"
+                onClick={handleOk}>
+                {submit.text}
+              </Button>
+            </div>
+          </div>
+        }
+        maskClosable={true}
+        open={modalIsOpen}
+        style={{
+          minWidth: size === 'small' ? 358 : size === 'medium' ? 692 : 1025,
+        }}
+        title={
+          <div style={{ display: 'flex' }}>
+            {danger ? (
+              <div className={css.dangerIcon} style={{ paddingRight: 16 }}>
+                <ExclamationCircleOutlined />
+              </div>
+            ) : icon ? (
+              <div style={{ paddingRight: 16 }}>
+                <Icon name={icon} />
+              </div>
+            ) : null}
+            <div style={{ paddingRight: 4 }}>{titleText}</div>
+            {headerLink && (
+              <Link path={headerLink.url} popout>
+                {headerLink.text}
+              </Link>
+            )}
+          </div>
+        }
+        onCancel={close}
+        onOk={handleOk}>
+        {modalBody}
```

Comment:
RE: 
> When the modal is submitting a form contained within should have its fields disabled

@johnkim-det something like replacing this with `<Spinner spinning={isSubmitting}>{modalBody}</Spinner>` could work.